Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -464,11 +470,80 @@ public CsharpCompiler(ErrorReporter reporter)
 
       CompileDatatypeDestructorsAndAddToInterface(dt, wr, interfaceTree, DtT_TypeArgs);
 
+      DtDowncastClone(dt, interfaceTree, nonGhostTypeArgs, toInterface: true);
+      if (!dt.IsRecordType) {
+        DtDowncastClone(dt, wr, nonGhostTypeArgs);
+      }
+
       CompileDatatypeInterfaceMembers(dt, interfaceTree);
 
       return new ClassWriter(this, btw, null);
     }
 
+    private void DtDowncastClone(DatatypeDecl dt, ConcreteSyntaxTree wr, List<TypeParameter> nonGhostTypeArgs, bool toInterface = false, bool lazy = false, DatatypeCtor ctor = null) {
+      if (nonGhostTypeArgs.Any(ty => ty.Variance == TypeParameter.TPVariance.Contra)
+          || dt.Ctors.Any(ctor => ctor.Formals.Any(f => f.Type.IsArrowType || f.Type.IsRefType))) {
+        return;
+      }
+      var typeArgs = TypeParameters(nonGhostTypeArgs, safe: true);
+      var dtArgs = "_I" + dt.CompileName + typeArgs;
+      string PrintConverter((TypeParameter, int) t) {
+        var (tArg, i) = t;
+        var name = IdName(tArg);
+        return $"Func<{name}, __{name}> converter{i}";
+      }
+      var argsWithIndex = nonGhostTypeArgs.Zip(Enumerable.Range(0, nonGhostTypeArgs.Count));
+
+      if (!toInterface) {
+        wr.Write($"public {(ctor != null || lazy ? (dt.IsRecordType ? "" : "override ") : "abstract ")}");
+      }
+
+      wr.Write($"{dtArgs} DowncastClone{typeArgs}({argsWithIndex.Comma(PrintConverter)})");
+
+      if (ctor == null && !lazy) {
+        wr.WriteLine(";");
+        return;
+      }
+
+      string PrintInvocation((string, Type) t) {
+        var (name, type) = t;
+        var constructorIndex = -1;
+        bool ContainsTyVar(TypeParameter tp, Type ty)
+          => (ty.AsTypeParameter != null && ty.AsTypeParameter.Equals(tp))
+             || ty.TypeArgs.Exists(ty => ContainsTyVar(tp, ty));
+
+        if ((constructorIndex = nonGhostTypeArgs.IndexOf(type.AsTypeParameter)) != -1) {
+          return $"converter{constructorIndex}({name})";
+        }
+
+        if (nonGhostTypeArgs.Exists(ty => ContainsTyVar(ty, type))) {
+          var map = nonGhostTypeArgs.ToDictionary(
+            tp => tp,
+            tp => (Type)new UserDefinedType(tp.tok, new TypeParameter(tp.tok, $"_{tp.Name}", tp.VarianceSyntax)));
+          var to = Resolver.SubstType(type, map);
+          var downcast = new ConcreteSyntaxTree();
+          EmitDowncast(type, to, null, downcast).Write(name);
+          return downcast.ToString();
+        }
+
+        return name;
+      }
+
+      var wBody = wr.NewBlock("").WriteLine($"if (this is {dtArgs} dt) {{ return dt; }}");
+      var parameter = "";
+
+      if (lazy) {
+        parameter = $"() => c().DowncastClone{typeArgs}({Enumerable.Range(0, nonGhostTypeArgs.Count).Comma(i => $"converter{i}")})";
+      } else {
+        var nonGhostFormals = ctor.Formals.FindAll(f => !f.IsGhost);
+        parameter = nonGhostFormals
+          .Zip(Enumerable.Range(0, nonGhostFormals.Count))
```

Comment:
Have you considered removing `Zip`, using the index supplying overload of `Comma`, and moving what Select is doing into `PrintInvocation`, so `PrintInvocation` takes a `Formal` and an index?